### PR TITLE
Reimplement `calc!`, and improve syntax support

### LINKED ISSUE #1088: `broadcast use _;` is not supported in the `calc!` macro
and it's often needed there.

cc @jaylorch

### PR DESCRIPTION
**Pull request by jaybosamiya**
This PR switches `calc!` away from the existing `macro_rules!`-based implementation (from #443) to a proc-macro-based implementation.

The primary motivation for this change is to improve maintainability, since the existing `macro_rules!`-based implementation can be a bit complicated to wrap one's head around. It needs to do a bunch of non-trivial indirection for supporting Verus syntax (some of it brittle), and is somewhat hard to extend (I had set the original implementation up to make it easy to add new relations, but if _say_ we wanted to add support for things like `by (bit_vector)` in the `macro_rules`-based implementation, it would be _quite_ hard).

In short, this implementation makes things more maintainable and extensible. It additionally makes it feasible for us to consider setting up `calc` as a keyword.

However, the current PR does none of the proposed extensions, and primarily is only a refactor, to keep reviewing easier. It _does_ fix up some existing issues, however; specifically, it now supports full Verus syntax inside the blocks (thus should fix #1088).

### CODE REVIEW

File: source/builtin_macros/src/calc_macro.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            input.parse::<Token![>]>()?;
+            Ok(Self::Gt)
+        } else {
+            Err(lookahead.error())
+        }
+    }
+}
+
+impl RelationOp {
+    /// Confirm that an `middle` relation is consistent with the main relation (self)
```
Comment (parno) [inline source/builtin_macros/src/calc_macro.rs:150]: "an" -> "a"

**Review by parno — APPROVED** [review]
Looks like a nice improvement -- this seems much easier to expand/adjust.

File: source/builtin_macros/src/calc_macro.rs (hunk 1 above)
Comment (jaybosamiya) [reply source/builtin_macros/src/calc_macro.rs]: Thanks for catching the typo, fixed!

File: source/builtin_macros/src/calc_macro.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+use syn_verus::token;
+use syn_verus::ExprBlock;
+use syn_verus::Token;
+use syn_verus::{parenthesized, Block, Expr};
+
+use crate::cfg_erase;
+use crate::syntax::rewrite_expr;
+
+fn rewrite_expr_to_token_stream(expr: &Expr) -> TokenStream {
+    TokenStream::from(rewrite_expr(cfg_erase(), true, expr.to_token_stream().into()))
```
[the remark this answers is not in the document]
Comment (jaybosamiya) [reply source/builtin_macros/src/calc_macro.rs:17]: The `calc!` macro doesn't work outside ghost/proof-mode in the first place:
```rs
use vstd::calc_macro::*;
use vstd::prelude::*;

verus! {

exec fn foo() {
    calc! {
        (<)
        1int; {}
        2;
    }
}

proof fn bar() {
    calc! {
        (<)
        1int; {}
        2;
    }
}

} // verus!

fn main() {}
```
gives
```console
$ verus x.rs
error: cannot use 'assert ... by' in exec mode
  --> x.rs:7:5
   |
7  | /     calc! {
8  | |         (<)
9  | |         1int; {}
10 | |         2;
11 | |     }
   | |_____^
   |
   = help: use a `proof` block
   = note: this error originates in the macro `::builtin_macros::calc_proc_macro` which comes from the expansion of the macro `calc` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 1 previous error
```

Even prior versions of `calc` would not work inside exec-mode code fwiw, so this PR leaves that unchanged.

I think if/when we consider a `calc` keyword, we might consider moving that up to automatically work in exec scope too, but for now, I think we should leave things as-is at least as part of this PR (i.e., require the `proof` block)

[the remark this answers is not in the document]
Comment (utaal) [reply source/builtin_macros/src/calc_macro.rs:17]: Sounds good. And it makes sense that you wouldn't be able to tell if we're in a non-exec context with a macro. That error message isn't perfect, but it's good enough to know what's wrong.

**Review by utaal — APPROVED** [review]

File: source/builtin_macros/src/calc_macro.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaybosamiya) [reply source/builtin_macros/src/calc_macro.rs:17]: Thanks!  I didn't see any easy way to improve the message, but it _does_ suggest the correct fix in its help, so hopefully if anyone hits it, they won't be _too_ confused.

**Comment by jaybosamiya** [comment]
@tjhance I'm merging this in since we're at 2 accepted reviews, but please let me know if you find that some part could've been implemented better. I'd added you as a reviewer since some of your macros are implemented in a somewhat similar style, so you might have helpful suggestions. Thanks!
